tools/integration-test/src/tests/client_upgrade.rs

```diff
@@ -0,0 +1,577 @@
+}
+
+#[test]
+fn test_invalid_client_upgrade() -> Result<(), Error> {
+    run_binary_chain_test(&InvalidClientUpgradeTest)
+}
+
+#[test]
+fn test_height_too_high_client_upgrade() -> Result<(), Error> {
+    run_binary_chain_test(&HeightTooHighClientUpgradeTest)
+}
+
+#[test]
+fn test_height_too_low_client_upgrade() -> Result<(), Error> {
+    run_binary_chain_test(&HeightTooLowClientUpgradeTest)
+}
+
+struct ClientUpgradeTest;
+
+impl TestOverrides for ClientUpgradeTest {
+    /// Update the genesis file in order to reduce the time required to upgrade the chain.
+    fn modify_genesis_file(&self, genesis: &mut serde_json::Value) -> Result<(), Error> {
+        use serde_json::Value;
+
+        let max_deposit_periods = genesis
+            .get_mut("app_state")
+            .and_then(|app_state| app_state.get_mut("gov"))
+            .and_then(|gov| gov.get_mut("deposit_params"))
+            .and_then(|deposit_params| deposit_params.as_object_mut());
+
+        if let Some(max_deposit_period) = max_deposit_periods {
+            max_deposit_period.insert(
+                "max_deposit_period".to_owned(),
+                Value::String(MAX_DEPOSIT_PERIOD.to_string()),
+            );
+        } else {
+            return Err(Error::generic(eyre!(
+                "failed to update max_deposit_period in genesis file"
+            )));
+        }
+
+        let voting_period = genesis
+            .get_mut("app_state")
+            .and_then(|app_state| app_state.get_mut("gov"))
+            .and_then(|gov| gov.get_mut("voting_params"))
+            .and_then(|voting_params| voting_params.as_object_mut());
+
+        if let Some(voting_period) = voting_period {
+            voting_period.insert(
+                "voting_period".to_owned(),
+                Value::String(VOTING_PERIOD.to_string()),
+            );
+            Ok(())
+        } else {
+            Err(Error::generic(eyre!(
+                "failed to update voting_period in genesis file"
+            )))
+        }
+    }
+}
+
+impl BinaryChainTest for ClientUpgradeTest {
+    fn run<
+        ChainA: ibc_test_framework::prelude::ChainHandle,
+        ChainB: ibc_test_framework::prelude::ChainHandle,
+    >(
+        &self,
+        _config: &ibc_test_framework::prelude::TestConfig,
+        _relayer: ibc_test_framework::prelude::RelayerDriver,
+        chains: ibc_test_framework::prelude::ConnectedChains<ChainA, ChainB>,
+    ) -> Result<(), ibc_test_framework::prelude::Error> {
+        let foreign_clients = chains.clone().foreign_clients;
+
+        let src_chain_config = chains.handle_b().config().unwrap();
+        let dst_chain_config = chains.handle_a().config().unwrap();
+
+        let upgraded_chain_id = chains.chain_id_a().0.clone();
+
+        let src_client_id = foreign_clients.client_id_b().0.clone();
+
+        // Create and send an chain upgrade proposal
+        let opts = UpgradePlanOptions {
+            dst_chain_config,
+            src_chain_config,
```

Comment:
What is the purpose to pass in the chain configs inside `UpgradePlanOptions`? I tried removing the fields, and it does not indicate that the configs are used anywhere.

Reply:
Indeed it seems these two fields are never used. I removed them in this commit https://github.com/informalsystems/hermes/pull/2795/commits/fde11207a6f03fb3475f1225973493d0c2860a38